rslib/src/scheduler/queue/mod.rs

```diff
@@ -242,6 +243,78 @@ impl Collection {
         Ok(())
     }
 
+    /// Update queues for a batch of processed cards.
+    /// This is more efficient than updating the queues one by one.
+    pub(crate) fn update_queues_for_processed_cards(
+        &mut self,
+        processed_card_ids: &HashSet<CardId>,
+    ) -> Result<()> {
+        let timing = self.timing_today()?;
+        if let Some(queues) = &mut self.state.card_queues {
+            // Remove all processed cards from the queues at once
+            if !processed_card_ids.is_empty() {
+                let mut removed_intraday_learning = 0;
+
+                // Remove from intraday learning queue
+                queues.intraday_learning.retain(|entry| {
+                    if processed_card_ids.contains(&entry.id) {
+                        removed_intraday_learning += 1;
+                        false
+                    } else {
+                        true
+                    }
+                });
+                queues.counts.learning = queues
+                    .counts
+                    .learning
+                    .saturating_sub(removed_intraday_learning);
+
+                let mut removed_new = 0;
+                let mut removed_review = 0;
+                let mut removed_interday_learning = 0;
+
+                // Remove from main queue
+                queues.main.retain(|entry| {
+                    if processed_card_ids.contains(&entry.id) {
+                        match entry.kind {
+                            MainQueueEntryKind::New => removed_new += 1,
+                            MainQueueEntryKind::Review => removed_review += 1,
+                            MainQueueEntryKind::InterdayLearning => removed_interday_learning += 1,
+                        }
+                        false
+                    } else {
+                        true
+                    }
+                });
+
+                // Update counts
+                queues.counts.new = queues.counts.new.saturating_sub(removed_new);
+                queues.counts.review = queues.counts.review.saturating_sub(removed_review);
+                queues.counts.learning = queues
+                    .counts
+                    .learning
+                    .saturating_sub(removed_interday_learning);
+
+                // Update learning cutoff
+                queues.update_learning_cutoff_and_count();
+
+                // Requeue any learning cards that need to be requeued
+                for &card_id in processed_card_ids {
+                    if let Some(card) = self.storage.get_card(card_id)? {
+                        if let Some(_requeued_learning) =
+                            queues.maybe_requeue_learning_card(&card, timing)
+                        {
+                            // We don't need to save undo information here since
+                            // this is a batch operation
+                            // and grade now can be undone as a normal op
+                        }
```

Comment:
Looking at this more carefully, I find myself wondering if you've actually accidentally taken the second approach I mentioned? `maybe_clear_study_queues_after_op()` does not cover the grade now case, so at the end of the operation, so the next queue operation after this one causes the queues to be rebuilt. If I'm right about that, all this code could be dropped.

Reply:
I didn't know `maybe_clear_study_queues_after_op()` before you mentioned it :-). I will check whether the code is useless.

Reply:
OK, I guess this info means I can remove the function `update_queues_for_processed_cards` safely?

```
[rslib/src/scheduler/queue/mod.rs:213:9] &op = OpChanges {
    op: GradeNow,
    changes: StateChanges {
        card: true,
        note: false,
        deck: true,
        tag: false,
        notetype: false,
        config: false,
        deck_config: false,
        mtime: true,
    },
}
[rslib/src/scheduler/queue/mod.rs:214:9] &op.requires_study_queue_rebuild() = true
```